File: python/151-200/200. Number of Islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        stack = []
        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    stack = [(i, j)]
                    while stack:
                        x,y = stack.pop(0)
                        grid[x][y] = "0"
                        if x - 1 >=0 and grid[x-1][y] == "1":
                            grid[x-1][y] = "0"
                            stack.append((x - 1, y))
                        if x + 1 < len(grid) and grid[x+1][y] == "1":
                            grid[x+1][y] = "0"
                            stack.append((x + 1, y))
                        if y-1 >= 0 and grid[x][y-1] == "1":
                            grid[x][y-1] = "0"
                            stack.append((x, y - 1))
                        if y + 1 < len(grid[0]) and grid[x][y+1] == "1":
                            grid[x][y+1] = "0"
                            stack.append((x, y + 1))
                    res += 1
        return res
            

File: cpp/151-200/200. Number of Islands.cc
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty())
            return 0;
        int res = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') {
                    bfs(grid, i, j);
                    res ++;
                }
            }
        }
        return res;
    }
    void bfs(vector<vector<char>> &grid, int i, int j) {
        if (i < 0 || i >= grid.size())
            return;
        if (j < 0 || j >= grid[0].size())
            return;
        queue<pair<int, int>> q;
        q.push(make_pair(i, j));
        while (!q.empty()) {
            pair<int, int> p = q.front();
            q.pop();
            int x = p.first, y = p.second;
            if(x < 0 || x >= grid.size() || y < 0 || y >= grid[0].size() || grid[x][y] == '0')
            	continue;
            grid[x][y] = '0';
            q.push(make_pair(x - 1, y));
            q.push(make_pair(x + 1, y));
            q.push(make_pair(x, y + 1));
            q.push(make_pair(x, y - 1));
        }
    }
};